src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -45,27 +46,37 @@ public void Initialize(Http2StreamContext context)
 
             _context = context;
 
-            _inputFlowControl = new StreamInputFlowControl(
-                context.StreamId,
-                context.FrameWriter,
-                context.ConnectionInputFlowControl,
-                context.ServerPeerSettings.InitialWindowSize,
-                context.ServerPeerSettings.InitialWindowSize / 2);
-
-            _outputFlowControl = new StreamOutputFlowControl(
-                context.ConnectionOutputFlowControl,
-                context.ClientPeerSettings.InitialWindowSize);
-
-            _http2Output = new Http2OutputProducer(
-                context.StreamId,
-                context.FrameWriter,
-                _outputFlowControl,
-                context.MemoryPool,
-                this,
-                context.ServiceContext.Log);
-
-            RequestBodyPipe = CreateRequestBodyPipe(context.ServerPeerSettings.InitialWindowSize);
-            Output = _http2Output;
+            // First time the stream is used we need to create flow control, producer and pipes.
+            // When a stream is reused these types will be reset and reused.
+            if (_inputFlowControl == null)
+            {
+                _inputFlowControl = new StreamInputFlowControl(
+                    this,
+                    context.FrameWriter,
+                    context.ConnectionInputFlowControl,
+                    context.ServerPeerSettings.InitialWindowSize,
+                    context.ServerPeerSettings.InitialWindowSize / 2);
+
+                _outputFlowControl = new StreamOutputFlowControl(
+                    context.ConnectionOutputFlowControl,
+                    context.ClientPeerSettings.InitialWindowSize);
```

Comment:
I'm concerned about pooling the StreamOutputFlowControl here. Couldn't the client update the InitialWindowSize for the connection and the stream wouldn't get the updated size as it wouldn't be updated on Reset? @Tratcher thoughts as well. I think _outputFlowControl should take an int for the InitialWindowSize for reset.

Reply:
Yeah, context.ClientPeerSettings.InitialWindowSize needs to be checked at the start of each new stream because it may have changed in between. That might just mean you need to pass the value into Reset.